ibis/mapd/client.py

```diff
@@ -0,0 +1,326 @@
+            result = pd.DataFrame(self.cursor.fetchall(), columns=col_names)
+        elif self.cursor is None:
+            result = pd.DataFrame([])
+        elif isinstance(self.cursor, pd.DataFrame):
+            result = self.cursor
+        elif GPUDataFrame is not None and isinstance(
+                self.cursor, GPUDataFrame
+        ):
+            result = self.cursor
+
+        return result
+
+    def __enter__(self):
+        # For compatibility when constructed from Query.execute()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        pass
+
+
+class MapDQuery(Query):
+    """
+
+    """
+    def _fetch(self, cursor):
+        # check if cursor is a pymapd cursor.Cursor
+        return self.schema().apply_to(cursor.to_df())
+
+
+class MapDClient(SQLClient):
+    """
+
+    """
+    database_class = Database
+    sync_query = MapDQuery
+    dialect = MapDDialect
+
+    def __init__(
+        self, uri: str=None, user: str=None, password: str=None,
+        host: str=None, port: int=9091, dbname: str=None,
+        protocol: str='binary', execution_type: int=3
+    ):
+        """
+
+        :param uri:
+        :param user:
+        :param password:
+        :param host:
+        :param port:
+        :param dbname:
+        :param protocol:
+        :param execution_type:
+        """
+        self.uri = uri
+        self.user = user
+        self.password = password
+        self.host = host
+        self.port = port
+        self.dbname = dbname
+        self.protocol = protocol
+        self.execution_type = execution_type
+
+        self.con = pymapd.connect(
+            uri=uri, user=user, password=password, host=host,
+            port=port, dbname=dbname, protocol=protocol
+        )
+
+    @property
+    def _table_expr_klass(self):
+        return ir.TableExpr
+
+    def log(self, msg):
+        log(msg)
+
+    def close(self):
+        """Close MapD connection and drop any temporary objects"""
+        self.con.close()
+
+    def _build_ast(self, expr, context):
+        """
+        Required.
+
+        :param expr:
+        :param context:
+        :return:
+        """
+        result = build_ast(expr, context)
+        return result
+
+    def _fully_qualified_name(self, name, database):
+        if bool(fully_qualified_re.search(name)):
+            return name
+
+        database = database or self.current_database
+        return '{0}.{1}'.format(database, name)
+
+    def _get_table_schema(self, table_name):
+        """
+
+        :param table_name:
+        :return:
+        """
+        database = None
+        table_name_ = table_name.split('.')
+        if len(table_name_) == 2:
+            database, table_name = table_name_
+        return self.get_schema(table_name, database)
+
+    def _execute(self, query, results=False):
+        """
+
+        :param query:
+        :return:
+        """
+        if not self.execution_type == 3:
```

Comment:
1. Don't negate `==` by using `not` at the beginning of the expression, so `not a == b` should be written `a != b`.
1. What does `3` mean here? Can you put it into some human readable enum?